src/sentry/static/sentry/app/stores/globalSelectionStore.jsx

```diff
@@ -1,143 +1,60 @@
+    this.listenTo(GlobalSelectionActions.initializeUrlState, this.onInitializeUrlState);
+    this.listenTo(GlobalSelectionActions.setOrganization, this.onSetOrganization);
     this.listenTo(GlobalSelectionActions.save, this.onSave);
     this.listenTo(GlobalSelectionActions.updateProjects, this.updateProjects);
     this.listenTo(GlobalSelectionActions.updateDateTime, this.updateDateTime);
     this.listenTo(GlobalSelectionActions.updateEnvironments, this.updateEnvironments);
   },
 
   reset(state) {
-    this._hasLoaded = false;
+    // Has passed the enforcement state
+    this._hasEnforcedProject = false;
+    this._hasInitialState = false;
     this.selection = state || getDefaultSelection();
   },
 
-  /**
-   * Initializes the global selection store
-   * If there are query params apply these, otherwise check local storage
-   */
-  loadInitialData(organization, queryParams, {api, skipLastUsed} = {}) {
-    this._hasLoaded = true;
-    this.organization = organization;
-    const query = pick(queryParams, Object.values(URL_PARAM));
-    const hasQuery = Object.keys(query).length > 0;
-
-    let globalSelection = getDefaultSelection();
-
-    if (hasQuery) {
-      const parsed = getStateFromQuery(queryParams);
-      globalSelection = {
-        projects: parsed.project || [],
-        environments: parsed.environment || [],
-        datetime: {
-          [DATE_TIME.START]: parsed.start || null,
-          [DATE_TIME.END]: parsed.end || null,
-          [DATE_TIME.PERIOD]: parsed.period || null,
-          [DATE_TIME.UTC]: parsed.utc || null,
-        },
-      };
-    } else if (!skipLastUsed) {
-      try {
-        const localStorageKey = `${LOCAL_STORAGE_KEY}:${organization.slug}`;
-
-        const storedValue = localStorage.getItem(localStorageKey);
-
-        const defaultDateTime = getDefaultSelection().datetime;
+  isReady() {
+    return this._hasInitialState;
+  },
 
-        if (storedValue) {
-          globalSelection = {datetime: defaultDateTime, ...JSON.parse(storedValue)};
-        }
-      } catch (ex) {
-        console.error(ex); // eslint-disable-line no-console
-        // use default if invalid
-      }
-    }
-    this.loadSelectionIfValid(globalSelection, organization, api);
+  onSetOrganization(organization) {
+    this.organization = organization;
   },
 
-  async loadSelectionIfValid(globalSelection, organization, api) {
-    if (await isValidSelection(globalSelection, organization, api)) {
-      this.selection = globalSelection;
-      this.trigger(this.selection);
-    }
+  /**
+   * Initializes the global selection store data
+   * Use query params if they exist, otherwise check local storage
+   */
+  onInitializeUrlState(newSelection) {
+    this._hasInitialState = true;
+    this.selection = newSelection;
+    this.trigger(this.get());
   },
 
   get() {
-    return this.selection;
+    return {
+      selection: this.selection,
+      isReady: this.isReady(),
+    };
```

Comment:
This is safe-ish since nothing accesses the store directly (instead it's accessed via `withGlobalSelection` HoC).